packages/graph/hash_graph/lib/graph/src/store/postgres/query/compile.rs

```diff
@@ -0,0 +1,307 @@
+    pub fn compile(&self) -> (String, &[&'f (dyn ToSql + Sync)]) {
+        struct Transpiler<'a, T>(&'a T);
+        impl<T: Transpile> Display for Transpiler<'_, T> {
+            fn fmt(&self, fmt: &mut Formatter<'_>) -> std::fmt::Result {
+                self.0.transpile(fmt)
+            }
+        }
+
+        (
+            Transpiler(&self.statement).to_string(),
+            &self.artifacts.parameters,
+        )
+    }
+
+    /// Compiles a [`Filter`] to a `Condition`.
+    pub fn compile_filter(&mut self, filter: &'f Filter<'q, T>) -> Condition<'q> {
+        if let Some(condition) = self.compile_special_filter(filter) {
+            return condition;
+        }
+
+        let condition = match filter {
+            Filter::All(filters) => Condition::All(
+                filters
+                    .iter()
+                    .map(|filter| self.compile_filter(filter))
+                    .collect(),
+            ),
+            Filter::Any(filters) => Condition::Any(
+                filters
+                    .iter()
+                    .map(|filter| self.compile_filter(filter))
+                    .collect(),
+            ),
+            Filter::Not(filter) => Condition::Not(Box::new(self.compile_filter(filter))),
+            Filter::Equal(lhs, rhs) => Condition::Equal(
+                lhs.as_ref()
+                    .map(|expression| self.compile_filter_expression(expression)),
+                rhs.as_ref()
+                    .map(|expression| self.compile_filter_expression(expression)),
+            ),
+            Filter::NotEqual(lhs, rhs) => Condition::NotEqual(
+                lhs.as_ref()
+                    .map(|expression| self.compile_filter_expression(expression)),
+                rhs.as_ref()
+                    .map(|expression| self.compile_filter_expression(expression)),
+            ),
+        };
+        condition
+    }
+
+    /// Compiles the `path` to a condition, which is searching for the latest version.
+    fn compile_latest_version_filter(
+        &mut self,
+        path: &'q T::Path<'q>,
+        operator: EqualityOperator,
+    ) -> Condition<'q> {
+        let mut version_column = Column {
+            table: Table {
+                name: path.terminating_table_name(),
+                alias: None,
+            },
+            access: path.column_access(),
+        };
+        // Depending on the table name of the path the partition table is selected
+        let partition_column = match version_column.table.name {
+            TableName::TypeIds => Column {
+                table: version_column.table,
+                access: ColumnAccess::Table { column: "base_uri" },
+            },
+            TableName::DataTypes => unreachable!(),
+        };
+
+        // Add a WITH expression selecting the partitioned version
+        self.statement
+            .with
+            .add_statement(version_column.table.name, SelectStatement {
```

Comment:
I will experiment with this a bit and try to find a better solution. This is a very valid point.

Reply:
I played around with this a little. Generally, it would be possible to rename `type_ids`, but then we would need to define it in the `TableName` enumeration (or change a fair bit of code to allow arbitrary tables). I think it's not worth changing it right now, but if we need to add more CTEs we can revisit this. I think it would be better to extend the CTE rather than add another one.

Reply:
Okay, thanks for checking this out. Maybe it would be good to capture somewhere as a potential issue we may have :+1:

Reply:
I added a warning in 5755c50a5f747a222329e5fc8bf9a5298d5749a5